### ISSUE TITLE
urwid LICENSE

### ISSUE THREAD

**Issue by tony** [urwid/urwid#41]
@wardi:

Is there any reason for the LGPL license on urwid? Would you consider changing it to a permissive license like MIT or BSD if there was a good reason?

**Comment by tony** [comment]
@wardl Where does this stand?

**Comment by wardi** [comment]
@tony someone or some service need to contact all the contributors to get them to sign off on the license change.

**Comment by tony** [comment]
@beberlei: Is http://licenses.beberlei.de/licenses/projects/create working now?

**Comment by nyov** [comment]
It's really nice to see the responses, especially so fast, and willingness of people to go with a permissive license (even though it's not everyone's cup of tea).
Great community, was worth my effort in writing the script :heart: 

If anyone cares about the script used to get the repo statistics, it's a bit fugly but here it is: https://gist.github.com/nyov/5d085ec49d4497bb97ba

**Comment by jeblair** [comment]
I certainly don't want to argue with tony about his feelings on licensing.  I'll just say that I am happy with urwid being licensed under the LGPL and I haven't seen any problems with that in practice (including its use with ASL licensed programs).

**Comment by nyov** [comment]
I've been on both sides of the fence myself.
In fact at times it's hard for me to prefer one over the other.

**Comment by nyov** [comment]
Half the contributors responded in under 6 hours, that is just awesome!

So far there is a chance this might be happening.
But the coin could quickly flip the other way, with a few heavy contributors still on the fence.

**Comment by jeblair** [comment]
Also, if you really would rather use a non-copyleft license, you may want to consider the Apache license in preference to MIT.  It, like the LGPL, has some patent protections that you would be giving up if you moved from LGPL to MIT.

**Comment by ssokolow** [comment]
But it's incompatible with the GPL 2.0 so, if you want to be compatible with "GPL 2.0 only" stuff, you need to license under "your choice of GPL 2.0 or later or Apache 2.0"

**Comment by horazont** [comment]
I agreed because I would also have contributed if the project had been under the MIT at that time. I view the transition critically though.

**Comment by tony** [comment]
This LGPL stuff also causes complications with other libraries, a few that
come to mind are https://github.com/PySide/PySide, pygame and
https://github.com/libgit2/pygit2.

Sometimes well-intentioned devs coming from a background in compiled
languages mistakenly believe the permissiveness passes into the world of
interpreted languages. This simply isn't the case, since code sharing in
python and node, for instance, is more complicated than the simple linking
of a header file.

I emailed the Python Foundation about how LGPL is interpreted with python,
the reality is LGPL hasn't been tested in court (even with a compiled
language) let alone an interpreted language.

> But it's incompatible with the GPL 2.0 so, if you want to be compatible
with "GPL 2.0 only" stuff, you need to license under "your choice of GPL
2.0 or later or Apache 2.0"

Is BSD with an optional patent exception (golang-style) compatible with GPL
2.0?

**Comment by ssokolow** [comment]
> Is BSD with an optional patent exception (golang-style) compatible with GPL 2.0?

I'm not sure. What GPL 3.0 and Apache 2.0 offer is a patent pledge (a guarantee that the contributor will not exercise any patents they hold which may cover the code they contribute).

Whether that conflicts with the GPL 2.0 in a golang-style BSD license will depend on whether it runs afoul of the GPL's "you may not impose any additional restrictions on people to whom you distribute this" clause.

**Comment by aszlig** [comment]
I'm disagreeing with this, because I really don't see a problem with LGPL, especially for Python libraries, but maybe I'm wrong.

Let's say I'm working for a company and want to include urwid code in my proprietary software. If I'm going to only distribute `.pyc` or `.pyo` files or use other means to compile and/or obfuscate the results, including (L)GPLed  source code is going to be a problem (which I'd certainly want as a free software author). Otherwise I just need to include the source code with the software I'm distributing.

And it won't be a problem at all if I'd only want to use the software in-house, it only applies if you *distribute* software, right?

Same applies for subclassing, as long as the code is available (usually it is anyway in an interpreted language).

**Comment by nyov** [comment]
So, while there are still outstanding responses, it's clear that even if they respond eventually, a relicensed release now would be somewhat crippled.

Whether to go through with it (and unpatching all the code) is something I'll leave others to decide and deal with.

> [aszlig] Otherwise I just need to include the source code with the software I'm distributing.

I'm not sure whether that is true or not, but I know a BSD licensed python project which has rejected including or depending on LGPL'd code. Maybe because it presents issues further downstream in a library?

**Comment by aszlig** [comment]
@nyov: Which project is it and what is the reasoning behind this?

**Comment by ssokolow** [comment]
I believe the issue is typically fear over how the courts would define "library" and "object file" in an interpreted language (even more worrying lately, given the way the Oracle API copyrightability battle has been going).

Maybe MPL 2.0 would be a suitable compromise if BSD is off the table? It's explicitly compatible with LGPL 2.1+, GPL 2.0+, and Apache 2.0+ and it bounds the copyleft-ness on a per-file basis rather than per-library, which is less ambiguous in a non-C/C++ language.

**Comment by ids1024** [comment]
> Otherwise I just need to include the source code with the software I'm distributing.

You have to include the source code *under the LGPL*. If LGPL code is mixed with code under a BSD, MIT, etc. license, the resulting code base would have to be LGPL. As stated before, it is unclear if subclassing would legally count, and thus require any program subclassing classes from urwid to be licenced under the LGPL.

**Comment by jeblair** [comment]
"Ian D. Scott" <notifications@github.com> writes:

>> Otherwise I just need to include the source code with the software I'm distributing.
>
> You have to include the source code *under the LGPL*. If LGPL code is
> mixed with code under a BSD, MIT, etc. license, the resulting code
> base would have to be LGPL. As stated before, it is unclear if
> subclassing would legally count, and thus require any program
> subclassing classes from urwid to be licenced under the LGPL.

Fortunately, I think that has been clarified in LGPL3.  From section 0:

  Defining a subclass of a class defined by the Library is deemed a mode
  of using an interface provided by the Library.

So an application using an LGPL3 library is in the same position whether
it subclasses from the library or not, which is to say, one can
distribute the LGPL3 library, even in binary form, with the application.

At least, that's my reading of the situation, though I am not a lawyer,
and this is not legal advice.

**Comment by ids1024** [comment]
@jeblair Well, that is good. Since urwid is LGPL 2.1 or any later version, that should make it fine for my current use and many others. (Though I too am not a lawyer, and won't guarantee that to anyone.) But @tony made some other good points in favor of a more permissive license earlier in this discussion, which are also worth considering, such as twisted going from LGPL to MIT due to how the license may apply to python code.

**Comment by garrison** [comment]
The main reason I disagree with the licensing change is that (from my perspective) the points given above seem to create unwarranted Fear, Uncertainty, and Doubt about the LGPL.  Surely each point raised is a legitimate question, but they should be reason to talk to a lawyer, not to try to relicense a project without having sought clarification.  I myself am not a lawyer, but I see no reason why the LGPL should be unclear as applied to Python libraries.  As @shackra mentioned above, the entire conversation here (perhaps until @aszlig's comment) seems pretty one-sided.  It seems that nobody has attempted to contact one of the pro-bono lawyers in this community to seek clarification.  (One example of such a law firm is the Software Freedom Law Center, where I was previously employed.)  The conversation years ago in the Twisted project seems to have had no clarification from lawyers either, as best I can tell.

**Comment by horazont** [comment]
@garrison I agree with you entirely.

Do you have any remaining contacts to the Software Freedom Law Center or similar organizations? This matter should be settled by a professional taking a look, if it is possible to settle it at all. If it is not possible to settle this matter in the general case, that would also be a valuable thing to know.

**Comment by nyov** [comment]
I feel this might go nowhere fast here. I believe everyone should agree to disagree. Different people have different views and preferences in Licenses, for whichever reasons (be they informed or misinformed). It is unlikely that we can resolve this by converting everyone to prefer one license over the other, when it hasn't happened so far -- both licenses are very widespread.

If we say for a second that everyone who agreed or disagreed to the MIT license was also expressing their preference of using such license, there is a clear vote. Of course that won't be so, some people will have decided for the benefit of the greater group and since I (badly maybe) formulated the request as a question of acceptance, not preference. But it's the only indicator to go on now (short of doing a full vote on preferred licenses which might do more harm than good by fracturing the consensus).

Being pragmatic, the MIT *does* allow relicensing, so it would be a possibility to have a MIT licensed codebase, and keeping a patchqueue with the LGPL contributions on top. People can keep using the software LGPL'd, as now, or use and contribute to the MIT codebase.
(Eventually one would be resolved by either people contributing LGPL only, outdating the MIT base, or by contributing MIT only, consuming the patchqueue. Though maintaining the patchqueue might not be appreciated.)

I agree that some reasoning in the initial posts may be thin and far-fetched at best, badmouthing the LGPL at worst; if someone is unsure whether they may use LGPL'd software somewhere, they should get legal advice, IMHO discussing here what is or is not legally possible and legal corner-cases will be open-ended and unproductive; only showcasing the uncertainties people might be troubled with when considering LGPL'd code.

I would like to think those contributors who voted, did so from their own views and beliefs, not the reasons given by the thread starter here, so shouldn't be considered invalid by arguing that one person's reasonings.

.

I personally support the standpoint that the software community is growing up, companies better understand the value of supporting free software for their own good (they get a better foundation to build better products on, faster, with community-proven and unified concepts).

So being more free and permissive and unencumbered doesn't detract from contributions much, or encourage stealing, for the most part.
But it frees people from headaches when considering license issues and reuseability, therefore is more a kind of mature gentlemans agreement.
(Wish we could get to the same point with this software patent uglyness. So I agree a license like Apache 2.0 might be preferrable.)

There will always be black sheep, intent on keeping code locked up, their own and yours if they can, for some IP protection reason or other.
But the same holds true for copyleft software - companies will reason that some GPL software
runs on their own cloud, inhouse, and modifications need not be opened up.
Or they might consider themself above law and disregard the license alltogether (wikipedia: gpl-violations.org), after all who would even know?

In contrast, consider possible values of re-use of free software in proprietary products (much as we hate them):
E.g. if BSD sockets had been GNU sockets, Microsoft et al couldn't have based the winsock API on it, and there would have been a vastly more different sockets beast in Windows. Or multiple incompatible stacks even still.
Does the convenience of a similar API for programmers, across OS', outweight "stealing" the concept? (One might argue if they hadn't, Windows could be less popular these days, and the *NIX world better off.)

And making proprietary code build on permissive/free software (and they will, because it's free) will eventually suck them in; the free software outbreeding their proprietary head. So in their own interest they will keep their closed-source "patchset" small and contribute to the free portions of the code, simply for a better TDC/TCO, or be left behind.
So free software will still prevail, not always necessary to be militant about it in your license.
In contrast, for a copyleft codebase, the company would (possibly have to) develop their own derivative, potentially multiple competitive versions by different vendors, fracturing the software landscape instead of unifying it. No gain for the copyleft codebase that is already around.
(Examples: OpenStack vs. AWS, GitHub and GitLab, ...)

**Comment by techtonik** [comment]
I am not a GPL fan, because I didn't understand all the implications of the word "Free" in time, but as a contributor I'd feel kind of betrayed if project switched license in the middle. For me the ethical solution could be to jump to `urwid2` or something even if I all for MIT and even CC-0 style of things with non-enforced crediting.

**Comment by tony** [comment]
I wrote the PSF about LGPL last year. Alluding to @garrison mentioning of having a lawyer view the particulars, I asked them if they'd fund an attorney. I don't want to quote them, but the conversation boiled down to there isn't caselaw either way.

@techtonik : SDL2 did something very similar.  They were LGPL and went ISC when version 2 came out.

https://www.youtube.com/watch?v=MeMPCSqQ-34&feature=youtu.be&t=435

I made this issue over 2 years ago and since moved on to different things.

I'm closing this issue, if you still want to change the license, feel free to open it.
